src/main/java/betterquesting/api2/storage/SimpleDatabase.java

```diff
@@ -1,21 +1,66 @@
 package betterquesting.api2.storage;
 
 import java.util.*;
-import java.util.Map.Entry;
 
 public abstract class SimpleDatabase<T> implements IDatabase<T>
 {
-    private final TreeMap<Integer, T> mapDB = new TreeMap<>();
-    
+    /**
+     * If the cache size would somehow exceed 24MB (on 64bit machines) we stop.
+     */
+    public static int CACHE_MAX_SIZE = 24 * 1024 * 1024 / 8;
+    /**
+     * If {@code mapDB.size < SPARSE_RATIO * (mapDB.lastKey() - mapDB.firstKey())} the database will be considered
+     * sparse and an cache array won't be built to save memory.
+     * <p>
+     * Under this sparsity a 10k element database will roughly result in a 0.5MB cache which is more than enough reasonable.
+     */
+    public static double SPARSE_RATIO = 0.15d;
+    /**
+     * When array cache is disabled, sort merge join will be used when query size is bigger than this
+     */
+    public static int SORT_MERGE_JOIN_THRESHOLD = 300;
+
+    final TreeMap<Integer, T> mapDB = new TreeMap<>();
+
     private final BitSet idMap = new BitSet();
-    private List<DBEntry<T>> refCache = null;
-    
+    private LookupLogicType type = null;
+    private LookupLogic<T> logic = null;
+
+    private LookupLogic<T> getLookupLogic()
+    {
+        if(type == null)
+        {
+            LookupLogicType newType = LookupLogicType.determine(this);
+            if(newType != type)
```

Comment:
no need to check this

Reply:
if you don't check this you rebuild the cache on every `bulkLookup` invocation. not wise.

Reply:
Wait a minute this seems to be code I forget to clean up. You are right here